yjit/src/backend/arm64/mod.rs

```diff
@@ -344,7 +344,14 @@ impl Assembler
                 Insn::Cmp { left, right } => {
                     let opnd0 = split_load_operand(asm, left);
                     let opnd0 = split_less_than_32_cmp(asm, opnd0);
-                    let opnd1 = split_shifted_immediate(asm, right);
+                    let split_right = split_shifted_immediate(asm, right);
+                    let opnd1 = match split_right {
+                        Opnd::InsnOut { .. } if opnd0.num_bits() != split_right.num_bits() => {
+                            split_right.with_num_bits(opnd0.num_bits().unwrap()).unwrap()
+                        },
+                        _ => split_right
+                    };
```

Comment:
What does this change do?

Reply:
In the case an immediate cannot be used in one instruction (for example `4097`) the backend will move it in to a register.  The problem is that it always moves it to a 64 bit register when we may have requested a comparison with a 32 bit register.  If the number of bits in the split register doesn't match the number of bits in the comparison register, then we convert the split register to match the other register.

I'm not 100% sure if this is the right solution, but the test below exercises this case, and we ran in to this problem when trying to do the 16 / 32 bit comparison for the shape id.

Reply:
Makes sense 👍